### PR TITLE
irmin-bench: Improve summary's PP

### PR DESCRIPTION
**Pull request by Ngoguey42**
As in last PR, `trace_stat_summary_pp.ml` is not worth reviewing in too much details.

```
-- setups --
Hostname        | am-c3-small-x86-01
Word Size       | 64 bits
Start Time      | 2021/05/04 21:17:06 (GMT)
Inode Config    | mls:32 bf:32 sh:256
Store Type      | pack
Path Conversion | none

-- main metrics --       | 
CPU time elapsed         |    39m03s
TZ-transactions per sec  |   185.299
TZ-operations per sec    |  1657.679
Context.set per sec      |  7962.130
tail latency (1)         |  12.121 s
                         | 
-- resource usage --     | 
disk IO (total)          | 
  IOPS (op/sec)          |   103_485
  throughput (bytes/sec) | 326.393 M
  total (bytes)          | 764.862 G
disk IO (read)           | 
  IOPS (op/sec)          |   103_104
  throughput (bytes/sec) | 289.558 M
  total (bytes)          | 678.542 G
disk IO (write)          | 
  IOPS (op/sec)          |       381
  throughput (bytes/sec) |  36.836 M
  total (bytes)          |  86.320 G
                         | 
max memory usage         |   2.326 G
mean CPU usage           |       95%

  - (1) Longest Context.commit.
  - The "per sec" stats are calculated over CPU time.
  - "TZ-transactions" and "TZ-operations" are approximations.
  - "max memory usage" is the max size of OCaml's major heap.
  - "mean CPU usage" is inexact.

-- curves --
Block played count *C                  | 0 (before) |   50000    |   100000   |   150000   | 200000 (end)
Blocks progress *C                     |     0%     |    25%     |    50%     |    75%     |     100%
                                       |            |            |            |            | 
Wall time elapsed *C                   |        0us |    4min11s |   12min58s |   24min58s |      41min2s
CPU time elapsed *C                    |        0us |    4min11s |   12min44s |    24min6s |      39min3s
                                       |            |            |            |            | 
Approx. TZ-transaction count *C        |          0 |     32_135 |     78_011 |    165_596 |      434_225
Approx. TZ-operations count *C         |          0 |    474_247 |  1_411_012 |  2_503_734 |    3_884_565
Op count *C                            |    0.000 M |   17.118 M |   38.872 M |   63.160 M |     92.734 M
Read only ops per sec *LA              |        n/a |  39045.558 |  29172.034 |  24506.503 |    25219.179
Read write ops per sec *LA             |        n/a |   9478.318 |   7968.366 |   6767.481 |     8948.607
                                       |            |            |            |            | 
Commit per sec *LA                     |        n/a |    123.555 |     78.616 |     63.417 |       29.906
Add per sec *LA                        |        n/a |   9210.253 |   7829.304 |   6574.326 |     8341.141
Remove per sec *LA                     |        n/a |    267.576 |    138.758 |    192.915 |      607.347
Find per sec *LA                       |        n/a |  29380.600 |  21848.493 |  17874.302 |    16235.220
Mem per sec *LA                        |        n/a |   5586.697 |   4729.196 |   4539.451 |     7997.051
Mem_tree per sec *LA                   |        n/a |   4078.261 |   2594.345 |   2092.750 |      986.908
Copy per sec *LA                       |        n/a |      0.489 |      0.305 |      0.240 |        0.119
                                       |            |            |            |            | 
Add count per block *LA                |        n/a |     74.544 |     99.589 |    103.669 |      278.911
Remove count per block *LA             |        n/a |      2.166 |      1.765 |      3.042 |       20.308
Find count per block *LA               |        n/a |    237.793 |    277.913 |    281.855 |      542.874
Mem count per block *LA                |        n/a |     45.216 |     60.155 |     71.581 |      267.406
Mem_tree count per block *LA           |        n/a |     33.008 |     33.000 |     33.000 |       33.000
Copy count per block *LA               |        n/a |      0.004 |      0.004 |      0.004 |        0.004
                                       |            |            |            |            | 
Block duration *LA                     |          0 |   8.094 ms |  12.720 ms |  15.769 ms |    33.438 ms
Buildup seen duration *LA              |          0 |   1.687 ms |   2.178 ms |   2.274 ms |     5.178 ms
Buildup unseen duration *LA            |        n/a |   0.680 ms |   0.727 ms |   0.711 ms |     2.621 ms
Commit duration *LA                    |        n/a |   5.726 ms |   9.815 ms |  12.783 ms |    25.638 ms
                                       |            |            |            |            | 
Add duration *LA                       |        n/a |   6.690 µs |   5.133 µs |   5.660 µs |     4.433 µs
Remove duration *LA                    |        n/a |   9.542 µs |   9.936 µs |   6.438 µs |     4.592 µs
Find duration *LA                      |        n/a |   4.349 µs |   5.016 µs |   5.165 µs |     5.958 µs
Mem duration *LA                       |        n/a |   2.135 µs |   3.509 µs |   2.302 µs |     2.106 µs
Mem_tree duration *LA                  |        n/a |   1.070 µs |   1.289 µs |   1.370 µs |     1.476 µs
Copy duration *LA                      |        n/a |   5.599 µs |   5.904 µs |   6.398 µs |    10.869 µs
Checkout duration *LA                  |        n/a |   1.600 µs |   1.727 µs |   2.024 µs |     2.245 µs
                                       |            |            |            |            | 
Disk bytes read *C                     |    0.000 G |   56.283 G |  209.339 G |  425.273 G |    678.542 G
Disk bytes written *C                  |    0.000 G |    2.176 G |   14.024 G |   39.802 G |     86.320 G
Disk bytes read per block *LA          |        n/a |  2_494_812 |  3_416_173 |  3_746_204 |    7_117_881
Disk bytes written per block *LA       |        n/a |    120_415 |    357_339 |    626_061 |    1_520_179
Disk bytes read per sec *LA *N         |        n/a |  308.247 M |  268.567 M |  237.572 M |    212.868 M
Disk bytes written per sec *LA *N      |        n/a |   14.878 M |   28.093 M |   39.703 M |     45.463 M
                                       |            |            |            |            | 
Disk read count *C                     |    0.000 M |   24.874 M |   83.842 M |  157.602 M |    241.612 M
Disk write count *C                    |         23 |    236_531 |    455_623 |    670_129 |      892_209
Disk read count per block *LA          |        n/a |    948.278 |   1396.944 |   1582.190 |     2015.933
Disk write count per block *LA         |        n/a |      4.501 |      4.238 |      4.258 |        5.042
Disk read count per sec *LA *N         |        n/a |    117_165 |    109_823 |    100_337 |       60_289
Disk write count per sec *LA *N        |        n/a |        556 |        333 |        270 |          151
                                       |            |            |            |            | 
pack.finds per block *LA               |        n/a |    712.510 |    961.787 |   1092.956 |     1639.277
pack.cache_misses per block *LA        |        n/a |     46.376 |     66.654 |     71.637 |      114.526
pack.appended_hashes per block *LA     |        n/a |  0.048_399 |  0.040_311 |  0.084_205 |    0.230_581
pack.appended_offsets per block *LA    |        n/a |    788.747 |   1159.751 |   1344.190 |     2249.057
                                       |            |            |            |            | 
tree.contents_hash per block *LA       |        n/a |     51.596 |     64.469 |     72.012 |      151.895
tree.contents_find per block *LA       |        n/a |    132.966 |    162.873 |    167.204 |      206.104
tree.contents_add per block *LA        |        n/a |     46.975 |     57.807 |     64.699 |      106.305
tree.node_hash per block *LA           |        n/a |    140.638 |    177.376 |    196.723 |      327.926
tree.node_mem per block *LA            |        n/a |    140.043 |    176.926 |    195.982 |      327.000
tree.node_add per block *LA            |        n/a |    138.036 |    173.827 |    192.548 |      324.165
tree.node_find per block *LA           |        n/a |    227.101 |    269.946 |    282.265 |      341.599
tree.node_val_v per block *LA          |        n/a |  5.289e-10 |  5.045e-16 |  4.811e-22 |    4.589e-28
tree.node_val_find per block *LA       |        n/a |          0 |          0 |          0 |            0
tree.node_val_list per block *LA       |        n/a |  1.567_878 |  1.482_379 |  2.284_458 |    1.760_123
                                       |            |            |            |            | 
index.nb_merge *C                      |          0 |         11 |         32 |         56 |           84
index.cumu_data_bytes *C               |    0.000 G |    1.241 G |   11.194 G |   34.751 G |     78.653 G
index.cumu_data_bytes per block *LA    |        n/a |     87_067 |    315_216 |    579_553 |    1_439_721
                                       |            |            |            |            | 
gc.minor_words allocated *C            |    0.001 G |   62.776 G |  178.178 G |  328.136 G |    522.771 G
gc.minor_words allocated per block *LA |        n/a |  1_859_625 |  2_727_972 |  3_256_813 |    6_875_605
gc.promoted_words *C                   |    0.000 G |    2.993 G |    8.073 G |   14.184 G |     21.679 G
gc.major_words allocated *C            |    0.001 G |    3.291 G |    9.911 G |   19.333 G |     32.773 G
gc.major_words allocated per block *LA |        n/a |    101_890 |    161_775 |    210_443 |      454_719
gc.minor_collections *C                |          4 |    239_591 |    680_223 |  1_252_904 |    1_997_210
gc.minor_collections per block *LA     |        n/a |      7.097 |     10.431 |     12.432 |       26.310
gc.major_collections *C                |          1 |        231 |        581 |      1_053 |        1_700
gc.major_collections per block *LA     |        n/a |  0.005_869 |  0.008_161 |  0.010_510 |    0.022_528
gc.compactions *C                      |          0 |          7 |         22 |         44 |           70
                                       |            |            |            |            | 
gc.major heap bytes top *C             |    7.172 M | 1599.009 M | 1804.403 M | 2326.278 M |   2326.278 M
gc.major heap bytes *LA                |    7.172 M | 1038.623 M | 1186.419 M |  759.869 M |   1329.029 M
                                       |            |            |            |            | 
index_data bytes *S                    |    0.000 M |  225.525 M |  699.599 M | 1241.697 M |   1871.698 M
index_data bytes per block *LA         |        n/a |      9_478 |     10_539 |     10_773 |       17_786
store_pack bytes *S                    |    0.000 M |  474.762 M | 1421.501 M | 2557.664 M |   3904.512 M
store_pack bytes per block *LA         |        n/a |     15_557 |     21_276 |     24_265 |       42_412
index_log bytes *S                     |    0.000 M |   22.500 M |   22.500 M |   22.500 M |     22.500 M
index_log_async *S                     |    0.000 M |    6.494 M |    9.791 M |   13.469 M |     21.423 M
store_dict bytes *S                    |         24 |  2_056_540 |  2_458_592 |  2_458_592 |    2_458_592
                                       |            |            |            |            | 
/data/contracts/index *S               |        n/a |          4 |          4 |          4 |            4
/data/big_maps/index *S                |        n/a |          0 |          0 |          0 |            0
/data/rolls/index *S                   |        n/a |        256 |        256 |        256 |          256
/data/rolls/owner/current *S           |        n/a |        256 |        256 |        256 |          256
/data/commitments *S                   |        n/a |        256 |        256 |        256 |          256
/data/contracts/index/ed25519 *S       |        n/a |        256 |        256 |        256 |          256
/data/contracts/index/originated *S    |        n/a |        256 |        256 |        256 |          256

Types of curves:
  *C: Cumulative. No smoothing.
  *LA: Local Average. Smoothed using a weighted sum of the value in the
       block and the exponentially decayed values of the previous blocks.
       Every 2500.01 blocks, half of the past is forgotten.
  *S: Size. E.g. directory entries, file bytes. No smoothing.
  *N: Very noisy.
```

### CODE REVIEW

File: bench/irmin-pack/trace_collection.ml [hunk 1]
```diff
@@ -66,34 +66,27 @@ module Make_stat (Store : Irmin.KV) = struct
...
+            | [] -> acc
+            | hd :: tl ->
+                if List.mem hd prev_merge_durations then (
+                  assert (List.length acc = 0) (* No oldie after a newies *);
+                  aux acc tl)
+                else aux ((hd /. 1e6) :: acc) tl
+          in
+          let l = aux [] v.merge_durations in
+          assert (List.length l > 0) (* At least one newie *);
+          l
```
Comment (Ngoguey42) [inline bench/irmin-pack/trace_collection.ml:89]: This fixes a bug

File: bench/irmin-pack/tree.ml [hunk 2]
```diff
@@ -919,6 +919,7 @@ let main_term =
     $ empty_blobs)
 
 let () =
+  (* TODO: Rename *)
```
Comment (Ngoguey42) [inline bench/irmin-pack/tree.ml:922]: @icristescu Did the `data.tarides.com` URLs change?

Comment (icristescu) [reply bench/irmin-pack/tree.ml:922]: I've updated them, the urls are
```
http://data.tarides.com/irmin/data4_10310commits.repr
http://data.tarides.com/irmin/data4_100066commits.repr
http://data.tarides.com/irmin/data_1343496commits.repr
```
let me know if they don't work.

Comment (Ngoguey42) [reply bench/irmin-pack/tree.ml:922]: Thanks

File: bench/irmin-pack/trace_collection.ml [hunk 3]
```diff
@@ -66,34 +66,27 @@ module Make_stat (Store : Irmin.KV) = struct
...
+          let rec aux acc = function
+            | [] -> acc
+            | hd :: tl ->
+                if List.mem hd prev_merge_durations then (
+                  assert (List.length acc = 0) (* No oldie after a newies *);
+                  aux acc tl)
+                else aux ((hd /. 1e6) :: acc) tl
+          in
+          let l = aux [] v.merge_durations in
+          assert (List.length l > 0) (* At least one newie *);
```
Comment (CraigFe) [inline bench/irmin-pack/trace_collection.ml:88]:
```suggestion
                  assert (acc = []) (* No oldie after a newies *);
                  aux acc tl)
                else aux ((hd /. 1e6) :: acc) tl
          in
          let l = aux [] v.merge_durations in
          assert (l <> []) (* At least one newie *);
```

Silly nit, but I dislike comparing `List.length l` to things for perf reasons.

File: bench/irmin-pack/trace_collection.ml [hunk 4]
```diff
@@ -66,34 +66,27 @@ module Make_stat (Store : Irmin.KV) = struct
...
-            assert (prev_nb_merge + List.length l = nb_merge);
-            l
-      in
-      let new_merge_durations =
-        List.map (fun v -> v /. 1e6) new_merge_durations
+        if v.merge_durations == prev_merge_durations then []
+        else
+          (* This is anoying to compute. We can't rely on nb_merge.
+             Assume that all merge durations are unique.
+             Assume that we never have >10 merges at one.
```
Comment (CraigFe) [inline bench/irmin-pack/trace_collection.ml:77]:
```suggestion
             Assume that we never have >10 merges at once.
```

File: bench/irmin-pack/trace_stat_summary_pp.ml [hunk 5]
```diff
@@ -56,420 +56,294 @@ end
...
-    `Data (`R, stat_name, curves)
-  in
-
-  let op_count : string -> Op.Key.t -> summary_floor =
-   fun name op ->
-    let curves =
-      zip (fun s -> Summary.(Op.Map.find op s.ops).count.evolution)
+    let tx_per_sec =
+      (* TODO: When replaying on a middle section of the chain, set
+         [~first_block_idx] *)
```
Comment (CraigFe) [inline bench/irmin-pack/trace_stat_summary_pp.ml:160]: Is this a feature we support?

File: bench/irmin-pack/trace_stat_summary_pp.ml [hunk 6]
```diff
@@ -505,31 +379,432 @@ let unsafe_pp sample_count ppf summary_names (summaries : Summary.t list) =
...
+     -- curves --\n\
      %s\n\n\
      Types of curves:\n\
     \  *C: Cumulative. No smoothing.\n\
     \  *LA: Local Average. Smoothed using a weighted sum of the value in the\n\
     \       block and the exponentially decayed values of the previous blocks.\n\
     \       Every %.2f blocks, half of the past is forgotten.\n\
-    \  *S: Size. E.g. directory entries, file bytes. No smoothing." tbl0 tbl1
+    \  *S: Size. E.g. directory entries, file bytes. No smoothing.\n\
+    \  *N: Very noisy." table0 table1 table2
```
Comment (CraigFe) [inline bench/irmin-pack/trace_stat_summary_pp.ml:807]: Style nit: this looks like it wants to be a multiline string `{| |}`

File: bench/irmin-pack/trace_stat_summary_utils.ml [hunk 7]
```diff
@@ -116,6 +116,246 @@ let pp_percent ppf v =
...
+(** List of tuples of week timestamp, week transaction count, week ops count and
+    week block count.
+
+    There is ~1 block per minute in tezos and 10080 minutes in a week. Hence the
+    content of the last columns.
+
+    Computed using:
+
+    https://api.tzstats.com/series/block.json?collapse=1w&columns=n_tx,n_ops,count *)
+let weekly_stats =
```
Comment (CraigFe) [inline bench/irmin-pack/trace_stat_summary_utils.ml:128]: Can we put this in a different file? e.g. `tezos_history_metrics.ml`. It's a bit odd to have Tezos-specific numbers in the main logic.

**Review by CraigFe — APPROVED** [review]
Mostly (all?) style nits, but LGTM overall.

File: bench/irmin-pack/trace_stat_summary_pp.ml (hunk 5 above)
Comment (Ngoguey42) [reply bench/irmin-pack/trace_stat_summary_pp.ml:160]: Not yet, but that's on the top of my TODO

File: bench/irmin-pack/trace_stat_summary_utils.ml (hunk 7 above)
Comment (Ngoguey42) [reply bench/irmin-pack/trace_stat_summary_utils.ml:128]: Sure

**Review by CraigFe — APPROVED** [review]

**Pull request merged**